articles/2019-03-25-le-type-option-c-est-quoi-et-ca-regle-quel-probleme.md

```diff
@@ -0,0 +1,177 @@
+                  /* ^ le compiler a compris qu'il s'agissait d'une chaîne de caractères! */
+```
+
+Cette généricité fait de l'option une abstraction générale pour représenter la présence ou l'absence de n'importe quel type de valeur. Cela nous permet par exemple de créer une fonction `map` :
+
+```reason
+let map = (opt, f) =>
+  switch (opt) {
+  | Some(x) => Some(f(x))
+  | None => None
+  };
+```
+
+Et cette fonction pourra être utilisée pour **n'importe quelle option**. Jetons un œil à sa signature :
+
+```reason
+let map: (option('a), 'a => 'b) => option('b);
+```
+
+On peut lire cette signature de cette façon :
+
+- on a une fonction map
+- elle prend une option contenant une valeur de type `a`
+- elle prend une fonction prenant une valeur de type `a` et retourne une valeur de type `b`
+- elle retourne une option contenant une valeur de type `b`
+
+```reason
+Some(2)->map(x => x * 3) // Some(6)
+None->map(x => x * 3) // None
+```
+
+Un autre exemple de fonction utile est `flatMap`:
+
+```reason
+let flatMap = (opt, f) =>
+  switch (opt) {
+  | Some(x) => f(x)
+  | None => None
+  };
+/* let flatMap: (option('a), 'a => option('b)) => option('b); */
+/* `get` retourne une option */
+get("profile")
+->flatMap(a => a->get("address"))
+->flatMap(b => b->get("zipCode"))
+```
+
+## Le problème résolu par le type option
+
+Prenons pour exemple la fonction `Array.prototype.find` de JavaScript :
+
+```js
+let result = array.find(item => item === undefined || item.active);
+```
+
+Cet code va retourner:
+
+- un objet s'il a un champ `active` ayant une valeur évaluée comme vraie
+- `undefined` si un item de `array` est `undefined`
+- `undefined` si rien n'est trouvé
+
+Avec cette implémentation naïve, nous sommes incapables de savoir dans quel cas nous nous trouvons : soit on a trouvé un item `undefined`, soit on a rien trouvé.
+
+_Notez que le problème se pose ici avec `undefined` mais qu'il en serait de même un tableau contenant des `null` et une fonction `find` d'une bibliothèque retournant `null` dans le cas où elle ne trouve rien_
+
+Si l'on veut être capable de faire la différence entre les deux derniers cas, on doit utiliser une autre fonction: `findIndex`:
+
+```js
+let index = array.findIndex(item => item === undefined || item.active);
+if (index == -1) {
+  // not found
+} else {
+  // found
+  let result = array[index];
+}
+```
+
+Le code est plus lourd, moins lisible, et manque d'expressivité. `find` ne nous donne pas assez d'information au travers de la valeur retournée: `undefined` est "applati", et requiert une logique supplémentaire (ici `index`, si un item est trouvé, il sera supérieur à `-1`, sinon il sera égal à `-1`)
+
+Le problème ne vient pas de la fonction `find` elle même mais de la façon dont `null` et `undefined` sont traités. `null` **est** la valeur, il la **remplace**. `option` **l'enrobe**: c'est un conteneur.
+
+```reason
+/* `getBy` est l'equivalent de `find` */
+let result = array->Belt.Array.getBy(
```

Comment:
Jusqu'alors le code était "assez" générique pour que je m'en foute du language utilisé.
Ici tu sort Belt.Array qui sort de nul part (tu parles pas de reasonml une fois dans le poste). Du coup soit tu précises que tes exemples sont en reason dans l'article (petite pub) soit tu ne le dis pas et dans ce cas tu imagine que t'as du pseudo code generique et tu fou "Array.find" en place de Belt.Array.getBy

Reply:
resolved in 0787fbdb